### PR TITLE
Muon - Correct Workspace Naming in New Algorithms

### LINKED ISSUE #22623: Muon - Naming conventions in grouping/pairing algorithms
For the naming conventions in recent algorithms #22452 #22510 to be consistent with the Muon Analysis GUI, they should correctly include summed/subtracted periods to the workspace name for multi period data. Currently this is not the case.

The properties **SummedPeriods** and **SubtractedPeriods** control this in the algorithms. By default the values are "1" and "" for single period data. An example name might then be;

`EMU0001111; Group; fwd; Counts; #1`

When using "1,2" for the **SummedPeriods** property the name should be modified to

`EMU0001111; Group; fwd; Counts; 1+2; #1`

or when using "2" for the **SubtractedPeriods** property the name should be modified to

`EMU0001111; Group; fwd; Counts; 1-2; #1`

### PR DESCRIPTION
**Pull request by TWJubb**
Closes #22623 , merge after #22619 is merged.

As per the linked issue, the naming conventions of the new muon algorithms `ApplyMuonDetectorGrouping` and `ApplyMuonDetectorGroupPairing` do not match the Muon Analysis GUI. This PR fixes that issue.

**Test 1**

- Open the Muon Aanlysis GUI (Interfaces->Muon->Muon Analysis), select the MUSR instrument. In the load run box give *15189* and hit enter. Data should appear in the ADS. Close the GUI.

- Open the `ApplyMuonDetectorGrouping` algorithm (in `Muon/Data Handling/`). Set the 

	- **InputWorkspace** : "MuonAnalysis"
	- **InputWorkspaceGroup** : "MUSR00015189" 
	- **GroupName** : "test"
	- **Grouping** : 1,2
	- **AnalysisType** : test both options (leaving **TimeMin** and **TimeMax** as defaults)
	- **RebinArgs** : leave blank
	- **TimeOffset** : 0.0

- Now run the algorithm, check that named workspaces are added to the workspace group (for the naming convention see the algorithm documentation). 

- Now change **SummedPeriods** to "1,2", check that the names are updated to include "; 1+2;". Now change the **SubtractedPeriods** to "2" and **SummedPeriods** to "1", the name should include "; 1-2;".

**Test 2**

- Follow the same instructions as above for the `ApplyMuonDetectorGroupPairing` algorithm, with the additional properties:

	- **Alpha** : 1.0
	- **SpecifyGroupsManually** : checked
	- **InputWorkspace** : MuonAnalysis
	- **Group1** and **Group2** : "1-5", "6-10"

Fixes #22623 

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
The release notes need to be moved from 3.13 to 3.14

File: Framework/Muon/src/MuonAlgorithmHelper.cpp [hunk 1]
```diff
@@ -220,6 +220,96 @@ void groupWorkspaces(const std::string &groupName,
...
+
+  std::vector<int> summedPeriodVec, subtractedPeriodVec;
+  const char delimeter(',');
+
+  std::vector<std::string> summedPeriodStrVec;
+  boost::split(summedPeriodStrVec, summedPeriods,
+               [delimeter](char c) { return c == delimeter; });
+  if (!summedPeriodStrVec[0].empty()) {
+    transform(summedPeriodStrVec.begin(), summedPeriodStrVec.end(),
+              back_inserter(summedPeriodVec), [](const std::string &p) {
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonAlgorithmHelper.cpp]: would `period` be a better name than `p`

File: Framework/Muon/src/MuonAlgorithmHelper.cpp [hunk 2]
```diff
@@ -220,6 +220,96 @@ void groupWorkspaces(const std::string &groupName,
...
+                }
+              });
+  }
+
+  std::vector<std::string> subtractedPeriodStrVec;
+  boost::split(subtractedPeriodStrVec, subtractedPeriods,
+               [delimeter](char c) { return c == delimeter; });
+  if (!subtractedPeriodStrVec[0].empty()) {
+    transform(subtractedPeriodStrVec.begin(), subtractedPeriodStrVec.end(),
+              back_inserter(subtractedPeriodVec), [](const std::string &p) {
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonAlgorithmHelper.cpp]: same as above

File: Framework/Muon/src/MuonAlgorithmHelper.cpp [hunk 3]
```diff
@@ -220,6 +220,96 @@ void groupWorkspaces(const std::string &groupName,
...
+                  return std::stoi(p);
+                } else {
+                  throw std::invalid_argument(
+                      "List contains a non-integer value : " + p);
+                }
+              });
+  }
+
+  std::vector<std::string> subtractedPeriodStrVec;
+  boost::split(subtractedPeriodStrVec, subtractedPeriods,
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonAlgorithmHelper.cpp]: as far as I can tell this is the same as the above except for the vector it populates and the vector that it process. Could we replace both with a function call?

**Comment by AnthonyLim23** [comment]
`applymuondetectorgrouppairing` falls over after loading HIFI 84447 into muon analysis.

**Comment by TWJubb** [comment]
> `applymuondetectorgrouppairing` falls over after loading HIFI 84447 into muon analysis.

I can't seem to reproduce the bug, when I load HIFI00084447 the algorithm opens ok

**Comment by AnthonyLim23** [comment]
I have just tried nightly and after opening MUSR62260 from Muon Analysis (I then deleted all but one workspace in the group), opening the `applyMuonDetectorGroupPair` caused mantid to crash (twice in a row now)

**Comment by TWJubb** [comment]
> I have just tried nightly and after opening MUSR62260 from Muon Analysis (I then deleted all but one workspace in the group), opening the `applyMuonDetectorGroupPair` caused mantid to crash (twice in a row now)

Again, I cannot seem to reproduce this bug on this branch. Perhaps this was a bug which is fixed by this PR? Especially if you are finding the bug on the nightly and I'm not seeing it on the branch.

**Comment by AnthonyLim23** [comment]
I just did a clean build and I see the crash on this branch

**Comment by AnthonyLim23** [comment]
I have only seen the bug on windows 10. If you are not using this for the loading then we have no need for this code and it should be deleted.

**Comment by TWJubb** [comment]
Deleted on request by @AnthonyLim23 as this code is no longer required by the new Muon interface.

**Pull request closed without merging**